File: apps/web/lib/api/lists.ts
import type { paths } from "@ims/api-types"

import { apiFetch } from "./client"

export type ListKind = "travel" | "shopping" | "collection" | "generic"

export type ListSummary =
  paths["/api/lists"]["get"]["responses"]["200"]["content"]["application/json"]["items"][number]

export type ListListResponse =
  paths["/api/lists"]["get"]["responses"]["200"]["content"]["application/json"]

export type ListDetail =
  paths["/api/lists/{list_id}"]["get"]["responses"]["200"]["content"]["application/json"]

export type ListEntry = ListDetail["entries"][number]

export type ListCreateBody =
  paths["/api/lists"]["post"]["requestBody"]["content"]["application/json"]

export type ListUpdateBody =
  paths["/api/lists/{list_id}"]["patch"]["requestBody"]["content"]["application/json"]

export type EntryCreateBody =
  paths["/api/lists/{list_id}/entries"]["post"]["requestBody"]["content"]["application/json"]

export type EntryUpdateBody =
  paths["/api/lists/{list_id}/entries/{entry_id}"]["patch"]["requestBody"]["content"]["application/json"]

export async function listLists(
  params: { kind?: ListKind; limit?: number; offset?: number },
  accessToken: string | null,
): Promise<ListListResponse> {
  const q = new URLSearchParams()
  if (params.kind) q.set("kind", params.kind)
  if (params.limit !== undefined) q.set("limit", String(params.limit))
  if (params.offset !== undefined) q.set("offset", String(params.offset))
  const res = await apiFetch(`/lists${q.size ? `?${q}` : ""}`, { accessToken })
  return (await res.json()) as ListListResponse
}

export async function createList(
  body: ListCreateBody,
  accessToken: string | null,
): Promise<ListSummary> {
  const res = await apiFetch("/lists", {
    method: "POST",
    accessToken,
    body: JSON.stringify(body),
  })
  return (await res.json()) as ListSummary
}

export async function getList(id: string, accessToken: string | null): Promise<ListDetail> {
  const res = await apiFetch(`/lists/${id}`, { accessToken })
  return (await res.json()) as ListDetail
}

export async function updateList(
  id: string,
  body: ListUpdateBody,
  accessToken: string | null,
): Promise<ListSummary> {
  const res = await apiFetch(`/lists/${id}`, {
    method: "PATCH",
    accessToken,
    body: JSON.stringify(body),
  })
  return (await res.json()) as ListSummary
}

export async function deleteList(id: string, accessToken: string | null): Promise<void> {
  await apiFetch(`/lists/${id}`, { method: "DELETE", accessToken })
}

export async function createEntry(
  listId: string,
  body: EntryCreateBody,
  accessToken: string | null,
): Promise<ListEntry> {
  const res = await apiFetch(`/lists/${listId}/entries`, {
    method: "POST",
    accessToken,
    body: JSON.stringify(body),
  })
  return (await res.json()) as ListEntry
}

export async function updateEntry(
  listId: string,
  entryId: string,
  body: EntryUpdateBody,
  accessToken: string | null,
): Promise<ListEntry> {
  const res = await apiFetch(`/lists/${listId}/entries/${entryId}`, {
    method: "PATCH",
    accessToken,
    body: JSON.stringify(body),
  })
  return (await res.json()) as ListEntry
}

export async function toggleEntry(
  listId: string,
  entryId: string,
  accessToken: string | null,
): Promise<ListEntry> {
  const res = await apiFetch(`/lists/${listId}/entries/${entryId}/toggle`, {
    method: "POST",
    accessToken,
  })
  return (await res.json()) as ListEntry
}

export async function deleteEntry(
  listId: string,
  entryId: string,
  accessToken: string | null,
): Promise<void> {
  await apiFetch(`/lists/${listId}/entries/${entryId}`, {
    method: "DELETE",
    accessToken,
  })
}

export async function reorderEntries(
  listId: string,
  entryIds: string[],
  accessToken: string | null,
): Promise<void> {
  await apiFetch(`/lists/${listId}/entries/reorder`, {
    method: "POST",
    accessToken,
    body: JSON.stringify({ entry_ids: entryIds }),
  })
}

File: apps/api/app/routes/lists.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.list import (
    ListCreate,
    ListDetail,
    ListEntryCreate,
    ListEntryRead,
    ListEntryUpdate,
    ListListResponse,
    ListSummary,
    ListUpdate,
    ReorderRequest,
)
from app.services import lists_service as svc

router = APIRouter(prefix="/api/lists", tags=["lists"])


@router.get("", response_model=ListListResponse)
async def list_lists(
    kind: str | None = None,
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ListListResponse:
    return await svc.list_lists(session, user.id, kind=kind, limit=limit, offset=offset)


@router.post("", response_model=ListSummary, status_code=status.HTTP_201_CREATED)
async def create_list(
    body: ListCreate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ListSummary:
    return await svc.create_list(session, user.id, body)


@router.get("/{list_id}", response_model=ListDetail)
async def get_list(
    list_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ListDetail:
    return await svc.get_list_detail(session, user.id, list_id)


@router.patch("/{list_id}", response_model=ListSummary)
async def update_list(
    list_id: UUID,
    body: ListUpdate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ListSummary:
    return await svc.update_list(session, user.id, list_id, body)


@router.delete(
    "/{list_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def delete_list(
    list_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> Response:
    await svc.delete_list(session, user.id, list_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/{list_id}/entries",
    response_model=ListEntryRead,
    status_code=status.HTTP_201_CREATED,
)
async def create_entry(
    list_id: UUID,
    body: ListEntryCreate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ListEntryRead:
    return await svc.create_entry(session, user.id, list_id, body)


@router.patch("/{list_id}/entries/{entry_id}", response_model=ListEntryRead)
async def update_entry(
    list_id: UUID,
    entry_id: UUID,
    body: ListEntryUpdate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ListEntryRead:
    return await svc.update_entry(session, user.id, list_id, entry_id, body)


@router.post("/{list_id}/entries/{entry_id}/toggle", response_model=ListEntryRead)
async def toggle_entry(
    list_id: UUID,
    entry_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ListEntryRead:
    return await svc.toggle_entry(session, user.id, list_id, entry_id)


@router.post(
    "/{list_id}/entries/reorder",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def reorder_entries(
    list_id: UUID,
    body: ReorderRequest,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> Response:
    await svc.reorder_entries(session, user.id, list_id, body)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.delete(
    "/{list_id}/entries/{entry_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def delete_entry(
    list_id: UUID,
    entry_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> Response:
    await svc.delete_entry(session, user.id, list_id, entry_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
